File: src/main/scala/com/kirkwoodwest/scalawoods/extensions/LaunchPadMiniExtension.scala
package com.kirkwoodwest.scalawoods.extensions

import com.bitwig.extension.callback.ShortMidiDataReceivedCallback
import com.bitwig.extension.controller.ControllerExtension
import com.bitwig.extension.controller.api.ControllerHost
import com.kirkwoodwest.scalawoods.trackbank.{CursorTrackBank, CursorTrackHelper}
import com.kirkwoodwest.scalawoods.utils.LogUtil
import com.kirkwoodwest.scalawoods.{DoubleGrid, HardwareBasic, HardwareLaunchPadMiniMK3}
import com.kirkwoodwest.scalawoods.utils.LogUtil.reportExtensionStatus

import scala.jdk.CollectionConverters.*
import java.util.HashMap

class LaunchPadMiniExtension(
                              val definition: LaunchPadMiniExtensionDefinition,
                              val host: ControllerHost)
  extends ControllerExtension(definition, host):
  var hardware: HardwareLaunchPadMiniMK3 = null
  override def init(): Unit =
    LogUtil.init(host)
    val midiCallbackFunc:ShortMidiDataReceivedCallback = (d:Int, d2:Int, d3: Int) => {
      host.println(s"Button Pressed: $d $d2 $d3")
    }

     hardware = HardwareLaunchPadMiniMK3(host, 0, midiCallbackFunc)
    //val hardwareBasic = HardwareBasic(host.getMidiInPort(0), host.getMidiOutPort(0))
    //hardwareBasic.setInputCallback(midiCallbackFunc)
    
    val cursorTrackHelper: CursorTrackHelper = CursorTrackHelper(host, 64)

    val cursorTrackBank:CursorTrackBank = CursorTrackBank(host, cursorTrackHelper, 16,0,4, "Clip Launcher")
    //Convert to list  cursorTrackBank.getCursors
    val cursorList =  cursorTrackBank.getCursors.asScala.toList

    DoubleGrid(host, cursorList, hardware)
    host.println("Initializing LaunchPad Mini Extension")

  end init

  override def exit(): Unit =
    // TODO: Perform any cleanup once the driver exits
    // For now just show a popup notification for verification that it is no longer running.
    reportExtensionStatus(this, "Exited")

  override def flush(): Unit =
    hardware.flush()

end LaunchPadMiniExtension

File: src/main/java/com/kirkwoodwest/extensions/LaunchPadMiniExtension.java
package com.kirkwoodwest.extensions;

import com.bitwig.extension.callback.ShortMidiDataReceivedCallback;
import com.bitwig.extension.controller.ControllerExtension;
import com.bitwig.extension.controller.api.ControllerHost;
import com.bitwig.extension.controller.api.CursorTrack;
import com.kirkwoodwest.launchpadminimk3.DoubleGrid;
import com.kirkwoodwest.launchpadminimk3.hardware.HardwareLaunchPadMiniMK3;
import com.kirkwoodwest.openwoods.trackbank.CursorTrackBank;
import com.kirkwoodwest.openwoods.trackbank.CursorTrackHelper;
import com.kirkwoodwest.openwoods.utils.LogUtil;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

public class LaunchPadMiniExtension extends ControllerExtension {
    private HardwareLaunchPadMiniMK3 hardware;

    protected LaunchPadMiniExtension(LaunchPadMiniExtensionDefinition definition, ControllerHost host) {
        super(definition, host);
    }

    @Override
    public void init() {
        LogUtil.init(getHost());
        ShortMidiDataReceivedCallback midiCallbackFunc = (d, d2, d3) -> {
            getHost().println("Button Pressed: " + d + " " + d2 + " " + d3);
        };

        hardware = new HardwareLaunchPadMiniMK3(getHost(), 0, midiCallbackFunc);

        CursorTrackHelper cursorTrackHelper = new CursorTrackHelper(getHost(), 64);

        CursorTrackBank cursorTrackBank = new CursorTrackBank(getHost(), cursorTrackHelper, 16, 0, 4, "Clip Launcher");

        List<CursorTrack> cursorList = StreamSupport.stream(cursorTrackBank.getCursors().spliterator(), false)
                .collect(Collectors.toList());

        new DoubleGrid(getHost(), "DoubleGrid", cursorList, hardware);

        getHost().println("Initializing LaunchPad Mini Extension");
    }

    @Override
    public void exit() {
        LogUtil.reportExtensionStatus(this, "Exited");
    }

    @Override
    public void flush() {
        hardware.flush();
    }
}
